src/metabase/models/card.clj

```diff
@@ -873,6 +874,62 @@ saved later when it is ready."
                (map? (:dataset_query card)) (update :dataset_query dissoc :lib/metadata))]
     (next-method card json-generator)))
 
+(defn- replaced-inner-query-for-native-card
+  [query {:keys [fields tables] :as _replacement-ids}]
+  (let [id->field           (if (empty? fields)
+                              {}
+                              (group-by :id
+                                        (t2/query {:select [[:f.id :id]
+                                                            [:f.name :column]
+                                                            [:t.name :table]
+                                                            [:t.schema :schema]]
+                                                   :from   [[:metabase_field :f]]
+                                                   :join   [[:metabase_table :t] [:= :f.table_id :t.id]]
+                                                   :where  [:in :f.id (set/union (into #{} (keys fields))
+                                                                                 (into #{} (vals fields)))]})))
+        id->table           (if (empty? tables)
+                              {}
+                              (group-by :id
+                                        (t2/query {:select [[:t.id :id]
+                                                            [:t.name :table]
+                                                            [:t.schema :schema]]
+                                                   :from   [[:metabase_table :t]]
+                                                   :where  [:in :t.id (set/union (into #{} (keys tables))
+                                                                                 (into #{} (vals tables)))]})))
+        remove-id           #(select-keys % [:column :table :schema])
+        column-only         :column ; I know this looks weird, but we want all the :table and :schema nonsense once
+                                    ; https://github.com/metabase/macaw/issues/32 lands
+        table-only          :table
+        get-or-throw-from   (fn [m] (fn [k] (if (contains? m k)
+                                              (get m k)
+                                              (throw (ex-info "ID not found" {:id k :available m})))))
+        column-replacements (-> fields
+                                (update-keys (comp column-only remove-id first (get-or-throw-from id->field)))
+                                (update-vals (comp column-only remove-id first (get-or-throw-from id->field))))
```

Comment:
better ways to DRY this up?

Reply:
well `index-by` will drop the `first`, `column-only` will be removed too, and... should be good enough?